src/join/join.jl

```diff
@@ -461,142 +511,334 @@ end
+        end
+    else
         ranges = Vector{UnitRange{T}}(undef, nrow(dsl))
-        if length(oncols_left) == 1 && nrow(dsr)>1
-            success, result = _join_left_dict(dsl, dsr, ranges, oncols_left, oncols_right, right_cols, Val(T); makeunique = makeunique, mapformats = mapformats, check = check, threads = threads, multiple_match = multiple_match, multiple_match_name = multiple_match_name, obs_id = obs_id, obs_id_name = obs_id_name)
+        if length(oncols_left) == 1 && type == :both && nrow(dsr)>1
+            success, result =  _join_inner_dict(dsl, dsr, ranges, oncols_left, oncols_right, right_cols, Val(T); makeunique = makeunique, mapformats = mapformats, check = check, threads = threads, multiple_match = multiple_match, multiple_match_name = multiple_match_name, obs_id = obs_id, obs_id_name = obs_id_name)
             if success
                 return result
             end
         end
-        idx, uniquemode = _find_permute_and_fill_range_for_join!(ranges, dsr, dsl, oncols_right, oncols_left, stable, alg, mapformats, accelerate; threads = threads)
+        idx, uniquemode = _find_permute_and_fill_range_for_join!(ranges, dsr, dsl, oncols_right, oncols_left, stable, alg, mapformats, accelerate && (onright_range == nothing || length(oncols_right)>1); nsfpaj = nsfpaj, threads = threads)
 
-        for j in 1:length(oncols_left)
-            _change_refpool_find_range_for_join!(ranges, dsl, dsr, idx, oncols_left, oncols_right, mapformats[1], mapformats[2], j; threads = threads)
+        for j in 1:length(oncols_left)-1
+            _change_refpool_find_range_for_join!(ranges, dsl, dsr, idx, oncols_left, oncols_right, mapformats[1], mapformats[2], j; nsfpaj = nsfpaj, threads = threads)
         end
+        _change_refpool_find_range_for_join!(ranges, dsl, dsr, idx, oncols_left, oncols_right, mapformats[1], mapformats[2], length(oncols_left); type = type, nsfpaj = nsfpaj, threads = threads)
     end
+    
     new_ends = map(x -> max(1, length(x)), ranges)
     our_cumsum!(new_ends)
     total_length = new_ends[end]
 
+    inbits = nothing
+    revised_ends = nothing
+    ra = nothing
+    if length(right_range_cols) == 2
+        inbits = zeros(Bool, total_length)
+        # TODO any optimisation is needed for pa?
+        _fl = identity
+        _fr = identity
+        if mapformats[1]
+            _fl = getformat(dsl, left_range_col)
+        end
+        if mapformats[2]
+            _fr = getformat(dsr, right_range_cols[2])
+        end
+        revised_ends = _mark_lt_part!(inbits, _columns(dsl)[left_range_col], _columns(dsr)[right_range_cols[2]], _fl, _fr, ranges, idx, new_ends, total_length < typemax(Int32) ? Val(Int32) : Val(Int64); strict = strict_inequality[2], threads = threads)   
+        inbit_ranges = Vector{UnitRange{T}}(undef, nrow(dsl))
+        ra = Vector{UnitRange{T}}(undef, nrow(dsl))
+        @_threadsfor threads for i in 1:length(ranges)
```

Comment:
put the core computations inside a separate function - it helps reducing allocations

Reply:
Not sure if we need this at all.